### PR TITLE
[MLIR][Analysis] Consolidate topological sort utilities 

### PR DESCRIPTION
**Pull request by Dinistro**
This PR attempts to consolidate the different topological sort utilities into one place. It adds them to the analysis folder because the `SliceAnalysis` uses some of these.

There are now two different sorting strategies: 
1. Sort only according to SSA use-def chains
2. Sort while taking regions into account. This requires a much more elaborate traversal and cannot be applied on graph regions that easily.

This additionally reimplements the region aware topological sorting because the previous implementation had an exponential space complexity.

I'm open to suggestions on how to combine this further or how to fuse the test passes.

### CODE REVIEW

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 1]
```diff
@@ -146,3 +151,93 @@ bool mlir::computeTopologicalSorting(
...
+  SetVector<Block *> sortedBlocks = getBlocksSortedByDominance(region);
+  for (Block *block : sortedBlocks) {
+    for (Operation &op : *block) {
+      if (toSort.contains(&op))
+        result.insert(&op);
+      for (Region &subRegion : op.getRegions()) {
+        // Skip regions that do not contain operations from `toSort`.
+        if (!relevantRegions.contains(&region))
+          continue;
+        topoSortRegion(subRegion, relevantRegions, toSort, result);
```
Comment (joker-eph) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:216]: Can you replace this recursion with an iterative algorithm please?
These recursions are ticking bombs...

File: mlir/include/mlir/Analysis/TopologicalSortUtils.h [hunk 2]
```diff
@@ -104,6 +104,14 @@ bool computeTopologicalSorting(
     MutableArrayRef<Operation *> ops,
     function_ref<bool(Value, Operation *)> isOperandReady = nullptr);
 
+/// Get a list of blocks that is sorted according to dominance. This sort is
+/// stable.
+SetVector<Block *> getBlocksSortedByDominance(Region &region);
+
+/// Sorts all operation in `toSort` topologically while also region semantics.
+/// Does not support multi-sets.
```
Comment (Moxinilian) [inline mlir/include/mlir/Analysis/TopologicalSortUtils.h:112]: Parse error! I think this sentence has a typo somewhere :)
Also `all operationS`.

Comment (Moxinilian) [reply mlir/include/mlir/Analysis/TopologicalSortUtils.h:112]: Ah, I'm being nitpicky today: could you also make the tense consistent with the comment above? Either `Get` and `Sort` or `Gets` and `Sorts`.

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 3]
```diff
@@ -146,3 +151,93 @@ bool mlir::computeTopologicalSorting(
...
+      }
+      current = current->getParentRegion();
+    }
+  }
+  auto firstRange = llvm::make_first_range(regionCounts);
+  traversedRegions.insert(firstRange.begin(), firstRange.end());
+  return res;
+}
+
+/// Topologically traverses `region` and insers all encountered operations in
```
Comment (Moxinilian) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:200]:
```suggestion
/// Topologically traverses `region` and inserts all encountered operations in
```

File: mlir/test/Dialect/Affine/slicing-utils.mlir [hunk 4]
```diff
@@ -202,15 +202,15 @@ func.func @slicing_test() {
...
-  // FWDBWD-NEXT: %[[v9:.*]] = "slicing-test-op"(%[[v7]], %[[v8]]) : (i7, i8) -> i9
+  // FWDBWD: %[[v1:.*]] = "slicing-test-op"() : () -> i1
+  // FWDBWD: %[[v2:.*]] = "slicing-test-op"() : () -> i2
+  // FWDBWD: %[[v3:.*]] = "slicing-test-op"() : () -> i3
+  // FWDBWD: %[[v4:.*]] = "slicing-test-op"() : () -> i4
+  // FWDBWD: %[[v5:.*]] = "slicing-test-op"(%[[v1]], %[[v2]]) : (i1, i2) -> i5
+  // FWDBWD: %[[v6:.*]] = "slicing-test-op"(%[[v3]], %[[v4]]) : (i3, i4) -> i6
+  // FWDBWD: %[[v7:.*]] = "slicing-test-op"(%[[v1]], %[[v5]]) : (i1, i5) -> i7
+  // FWDBWD: %[[v8:.*]] = "slicing-test-op"(%[[v5]], %[[v6]]) : (i5, i6) -> i8
+  // FWDBWD: %[[v9:.*]] = "slicing-test-op"(%[[v7]], %[[v8]]) : (i7, i8) -> i9
```
Comment (Moxinilian) [inline mlir/test/Dialect/Affine/slicing-utils.mlir:213]: Why did you remove the DAG-based filecheck in those tests?

**Review by Moxinilian — COMMENTED** [review]
Should be good to me once this and Mehdi's comment is addressed.

File: mlir/test/lib/Analysis/TestSlice.cpp [hunk 5]
```diff
@@ -26,20 +28,16 @@ struct TestTopologicalSortPass
...
-    llvm::errs() << "Testing : " << getOperation().getName() << "\n";
-    for (Operation *op : sortedOp) {
-      op->print(llvm::errs());
-      llvm::errs() << "\n";
-    }
+
+    auto i32Type = IntegerType::get(&getContext(), 32);
+    SetVector<Operation *> sortedOps = topologicalSort(toSort);
+    for (auto [index, op] : llvm::enumerate(sortedOps))
+      op->setAttr(kOrderIndex, IntegerAttr::get(i32Type, index));
```
Comment (Moxinilian) [inline mlir/test/lib/Analysis/TestSlice.cpp:40]: Okay, technically this changes the expected behavior of this pass as the description explicitly mentions printing. Could you update it so it talks about setting the ordering attributes instead?

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 6]
```diff
@@ -146,3 +151,93 @@ bool mlir::computeTopologicalSorting(
...
+  }
+  assert(blocks.size() == region.getBlocks().size() &&
+         "some blocks are not sorted");
+
+  return blocks;
+}
+
+/// Computes the common ancestor region of all operations in `ops`. Remembers
+/// all the traversed regions in `traversedRegions`.
+static Region *findCommonParentRegion(const SetVector<Operation *> &ops,
```
Comment (gysit) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:173]:
```suggestion
static Region *findCommonAncestorRegion(const SetVector<Operation *> &ops,
```
nit: I would rename to findCommonAncestorRegion.

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 7]
```diff
@@ -146,3 +151,93 @@ bool mlir::computeTopologicalSorting(
...
+      blocks.insert(traversal.begin(), traversal.end());
+    }
+  }
+  assert(blocks.size() == region.getBlocks().size() &&
+         "some blocks are not sorted");
+
+  return blocks;
+}
+
+/// Computes the common ancestor region of all operations in `ops`. Remembers
```
Comment (gysit) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:171]:
```suggestion
/// Computes the closest common ancestor region of all operations in `ops`. Remembers
```
nit: maybe also say the function returns nullptr if no common region is found?

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 8]
```diff
@@ -146,3 +151,93 @@ bool mlir::computeTopologicalSorting(
...
+/// all the traversed regions in `traversedRegions`.
+static Region *findCommonParentRegion(const SetVector<Operation *> &ops,
+                                      DenseSet<Region *> &traversedRegions) {
+  // Map to count the number of times a region was encountered.
+  llvm::DenseMap<Region *, size_t> regionCounts;
```
Comment (gysit) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:176]:
```suggestion
  DenseMap<Region *, size_t> regionCounts;
```
ultra nit:

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 9]
```diff
@@ -146,3 +151,93 @@ bool mlir::computeTopologicalSorting(
...
+  // Walk the region tree for each operation towards the root and add to the
+  // region count.
+  Region *res = nullptr;
+  for (Operation *op : ops) {
+    Region *current = op->getParentRegion();
+    while (current) {
+      // Insert or get the count.
+      auto it = regionCounts.try_emplace(current, 0).first;
+      size_t count = ++it->getSecond();
+      if (count == expectedCount) {
```
Comment (gysit) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:188]:
```suggestion
      if (regionCounts[current]++ == expectedCount) {
```
nit: Wouldn't operator[] be more readable here?

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 10]
```diff
@@ -146,3 +151,93 @@ bool mlir::computeTopologicalSorting(
...
+  }
+}
+
+SetVector<Operation *>
+mlir::topologicalSort(const SetVector<Operation *> &toSort) {
+  if (toSort.size() <= 1)
+    return toSort;
+
+  assert(llvm::all_of(toSort,
+                      [&](Operation *op) { return toSort.count(op) == 1; }) &&
```
Comment (gysit) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:228]: This should be guaranteed since `toSort` is a SetVector?

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 11]
```diff
@@ -146,3 +151,93 @@ bool mlir::computeTopologicalSorting(
...
+         "expected only unique set entries");
+
+  // First, find the root region to start the recursive traversal through the
+  // IR.
+  DenseSet<Region *> relevantRegions;
```
Comment (gysit) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:233]:
```suggestion
  DenseSet<Region *> ancestorRegions;
```
or maybe also ancestorRegionsToSort could be an alternative name?

**Review by gysit — COMMENTED** [review]
Thanks for fixing this!

File: mlir/lib/Analysis/TopologicalSortUtils.cpp (hunk 10 above)
Comment (Dinistro) [reply mlir/lib/Analysis/TopologicalSortUtils.cpp:228]: That was there before, but you are right 🙂

File: mlir/test/Dialect/Affine/slicing-utils.mlir (hunk 4 above)
Comment (Dinistro) [reply mlir/test/Dialect/Affine/slicing-utils.mlir:213]: It only used `CHECK-DAG` for some random sub-parts. Now that the underlying implementation is stricter, these checks break.

Sadly, CHECK-DAG cannot be used nicely to check if an output satisfies a topological order, due to reordering of subtrees.

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 12]
```diff
@@ -146,3 +151,135 @@ bool mlir::computeTopologicalSorting(
...
+
+    // Walk the region tree for each operation towards the root and add to the
+    // region count.
+    Region *res = nullptr;
+    for (Operation *op : toSort) {
+      Region *current = op->getParentRegion();
+      // Store the block as an ancestor block.
+      ancestorBlocks.insert(op->getBlock());
+      while (current) {
+
```
Comment (gysit) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:215]:
```suggestion
```
nit: drop strange newline.

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 13]
```diff
@@ -146,3 +151,135 @@ bool mlir::computeTopologicalSorting(
...
+    while (!stack.empty()) {
+      StackT current = stack.pop_back_val();
+      if (auto *region = dyn_cast<Region *>(current)) {
+        // A region's blocks need to be traversed in dominance order.
+        SetVector<Block *> sortedBlocks = getBlocksSortedByDominance(*region);
+        for (Block *block : llvm::reverse(sortedBlocks))
+          // Only add blocks to the stack that are ancestors of the operations
+          // to sort.
+          if (ancestorBlocks.contains(block))
+            stack.push_back(block);
```
Comment (gysit) [inline mlir/lib/Analysis/TopologicalSortUtils.cpp:250]:
```suggestion
        for (Block *block : llvm::reverse(sortedBlocks)) {
          // Only add blocks to the stack that are ancestors of the operations
          // to sort.
          if (ancestorBlocks.contains(block))
            stack.push_back(block);
        }
```

**Review by gysit — COMMENTED** [review]
Nice! 

Looks good modulo nits.

**Review by joker-eph — COMMENTED** [review]
LG with other folks comments' addressed. Thanks!

File: mlir/lib/Analysis/TopologicalSortUtils.cpp [hunk 14]
```diff
@@ -146,3 +151,135 @@ bool mlir::computeTopologicalSorting(
...
+
+namespace {
+class TopoSortHelper {
+public:
+  explicit TopoSortHelper(const SetVector<Operation *> &toSort)
+      : toSort(toSort) {}
+
+  /// Executes the topological sort of the operations this instance was
+  /// constructed with. This function will destroy the internal state of the
+  /// instance.
```
[the remark this answers is not in the document]
Comment (Dinistro) [reply mlir/lib/Analysis/TopologicalSortUtils.cpp:179]: This is a purely internal class that is not accessible from the outside. I therefore decided to not add clearing logic.

**Review by gysit — APPROVED** [review]
LGTM
